internal/controller/management/serviceaccounts/serviceaccounts.go

```diff
@@ -0,0 +1,301 @@
+	)
+	logger.Info("starting to create RoleBindings for all Projects")
+
+	projectList := &kargoapi.ProjectList{}
+	if err := r.listProjectFn(ctx, projectList); err != nil {
+		return fmt.Errorf("error listing Projects: %w", err)
+	}
+
+	// Loop through each Project to create or update the corresponding RoleBinding.
+	for _, project := range projectList.Items {
+
+		roleBinding := &rbacv1.RoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      roleBindingName,
+				Namespace: project.Name,
+			},
+			RoleRef: rbacv1.RoleRef{
+				APIGroup: rbacv1.GroupName,
+				Kind:     "ClusterRole",
+				Name:     "kargo-controller-secrets-readonly",
+			},
+			Subjects: []rbacv1.Subject{
+				{
+					Kind:      "ServiceAccount",
+					Name:      sa.Name,
+					Namespace: r.cfg.KargoNamespace,
+				},
+			},
+		}
+
+		// Try to create RoleBinding directly, update if it already exists.
+		if err := r.createRoleBindingFn(ctx, roleBinding); err != nil {
+			if !kubeerr.IsAlreadyExists(err) {
+				return fmt.Errorf("error creating RoleBinding %q for ServiceAccount %q in Project namespace %q: %w",
+					roleBinding.Name, sa.Name, project.Name, err)
+			}
+			if updateErr := r.updateRoleBindingFn(ctx, roleBinding); updateErr != nil {
+				return fmt.Errorf("error updating existing RoleBinding %q in Project namespace %q: %w",
+					roleBinding.Name, project.Name, updateErr)
+			}
+			logger.Debug("Updated existing RoleBinding for ServiceAccount", "roleBinding",
+				roleBindingName, "namespace", project.Name)
+		} else {
+			logger.Info("Created RoleBinding for ServiceAccount", "roleBinding", roleBindingName, "namespace", project.Name)
+		}
+	}
+	logger.Info("Completed creating RoleBindings for all Projects")
+	return nil
+}
+
+// removeControllerPermissions removes RoleBindings for the specified ServiceAccount
+// from all Projects. This function is called when the ServiceAccount is no longer
+// managed by the controller, ensuring that permissions are cleaned up to prevent
+// unauthorized access to resources.
+func (r *reconciler) removeControllerPermissions(ctx context.Context, sa types.NamespacedName) error {
+
+	roleBindingName := fmt.Sprintf("%s-readonly-secrets", sa.Name)
+
+	logger := logging.LoggerFromContext(ctx).WithValues(
+		"serviceaccount", sa.Name,
+		"roleBinding", roleBindingName,
+	)
+	logger.Info("Starting to delete RoleBindings for all Projects")
+
+	projectList := &kargoapi.ProjectList{}
+	if err := r.listProjectFn(ctx, projectList); err != nil {
+		return fmt.Errorf("error listing Projects: %w", err)
+	}
+
+	for _, project := range projectList.Items {
+
+		roleBinding := &rbacv1.RoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      roleBindingName,
+				Namespace: project.Name,
+			},
+		}
+
+		if err := r.deleteRoleBindingFn(ctx, roleBinding); err != nil {
+			if kubeerr.IsNotFound(err) {
+				logger.Debug("RoleBinding not found, skipping deletion", "roleBinding",
+					roleBindingName, "namespace", project.Namespace)
```

Comment:
I also suggest, a little higher up, around line 272, doing this:

```
projectLogger := logger.WithValues("namespace", project.Name)
```

Then use that throughout the rest of the loop so you get namespace info for free in every message you write with that logger.

Also... `project.Namespace` will always be `""` because Projects themselves are cluster-scoped resources. `project.Name` however, will always match the namespace associated with the Project.